addons/mail/models/discuss/ir_binary.py

```diff
@@ -0,0 +1,18 @@
+from odoo import models, _
+from odoo.exceptions import AccessError
+
+
+class IrBinary(models.AbstractModel):
+    _inherit = 'ir.binary'
+
+    def _find_record_check_access(self, record, access_token, field=None):
+        if record._name == "res.partner" and field == "avatar_128":
+            guest = self.env["mail.guest"]._get_guest_from_context()
+            if not guest or guest.sudo().channel_ids & record.sudo().channel_ids:
+                return record.sudo()
+            raise AccessError(_('Avatar access error'))
+        if record._name == "mail.guest" and field == "avatar_128":
+            if self.env.user.partner_id.channel_ids & record.sudo().channel_ids:
+                return record.sudo()
+            raise AccessError(_('Avatar access error'))
```

Comment:
Conditions look incorrect.
Top level should be `record._name in ["res.partner", "mail.guest"] and field == "avatar_128"` (only one condition)
Second level should be "if guest and public" check guest channels "else" check current partner channels.

"not guest return sudo" is definitely not correct.